rust/src/test.rs

```diff
@@ -0,0 +1,209 @@
+use std::{
+    borrow::BorrowMut,
+    sync::{Arc, Mutex},
+};
+
+use crate::{
+    constant, ctx,
+    run::aggregation::{
+        agg_circuit_keygen, agg_circuit_run, keccak_circuit_run,
+    },
+    run::inner::{keygen, mock, prove, run},
+    scaffold::{AxiomCircuitScaffold, RawCircuitInput},
+    subquery::{caller::SubqueryCaller, header::HeaderField, types::AssignedHeaderSubquery},
+    types::AxiomCircuitParams,
+    utils::get_provider,
+    witness,
+};
+use axiom_codec::HiLo;
+use axiom_eth::{
+    halo2_base::{
+        gates::{
+            circuit::BaseCircuitParams, GateChip, GateInstructions, RangeChip, RangeInstructions,
+        },
+        safe_types::SafeTypeChip,
+        AssignedValue, Context,
+    },
+    halo2curves::bn256::Fr,
+    keccak::promise::KeccakFixLenCall,
+    rlc::circuit::{builder::RlcCircuitBuilder, RlcCircuitParams},
+    snark_verifier_sdk::halo2::aggregation::AggregationConfigParams,
+    utils::keccak::decorator::RlcKeccakCircuitParams,
+    Field,
+};
+use ethers::providers::JsonRpcClient;
+
+#[derive(Debug, Clone, Default)]
+struct MyCircuitInput {
+    a: u64,
+    b: u64,
+}
+
+#[derive(Debug, Clone)]
+struct MyCircuitVirtualInput<F: Field> {
+    a: AssignedValue<F>,
+    b: AssignedValue<F>,
+}
+
+impl RawCircuitInput<Fr, MyCircuitVirtualInput<Fr>> for MyCircuitInput {
+    fn assign(&self, ctx: &mut Context<Fr>) -> MyCircuitVirtualInput<Fr> {
+        let a = ctx.load_witness(Fr::from(self.a));
+        let b = ctx.load_witness(Fr::from(self.b));
+        MyCircuitVirtualInput { a, b }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+struct MyCircuit;
+impl<P: JsonRpcClient> AxiomCircuitScaffold<P, Fr> for MyCircuit {
+    type CircuitInput = MyCircuitInput;
+    type VirtualCircuitInput = MyCircuitVirtualInput<Fr>;
+
+    fn virtual_assign_phase0(
+        &self,
+        builder: &mut RlcCircuitBuilder<Fr>,
+        range: &RangeChip<Fr>,
+        subquery_caller: Arc<Mutex<SubqueryCaller<P, Fr>>>,
+        callback: &mut Vec<HiLo<AssignedValue<Fr>>>,
+        inputs: Self::VirtualCircuitInput,
+    ) {
+        let gate = GateChip::<Fr>::new();
+        gate.add(ctx!(builder, 0), inputs.a, inputs.b);
```

Comment:
the macro is nice to have, but I think when using `ctx` a bunch, it's still better if you do:
```
let ctx = builder.base.main(0);
```
I don't think there should be problems in this case because you don't borrow the builder anywhere else. Let me know if it still has BorrowMut runtime error - this should be fixable.